crates/red_knot_test/src/matcher.rs

```diff
@@ -0,0 +1,731 @@
+    }
+}
+
+fn unmatched<'a, T: Unmatched + 'a>(unmatched: impl Into<&'a [T]>) -> Vec<String> {
+    unmatched.into().iter().map(Unmatched::unmatched).collect()
+}
+
+struct Matcher {
+    line_index: LineIndex,
+    source: SourceText,
+}
+
+impl Matcher {
+    fn from_file(db: &TestDb, file: File) -> Self {
+        Self {
+            line_index: line_index(db, file),
+            source: source_text(db, file),
+        }
+    }
+
+    /// Check a slice of [`Diagnostic`]s against a slice of [`Assertion`]s.
+    ///
+    /// Return vector of [`Unmatched`] for any unmatched diagnostics or assertions.
+    fn match_line<'a, 'b, T: Diagnostic + 'a>(
+        &self,
+        diagnostics: impl Into<&'a [T]>,
+        assertions: impl Into<&'a [Assertion<'b>]>,
+    ) -> Result<(), Vec<String>>
+    where
+        'b: 'a,
+    {
+        let mut failures = vec![];
+        let mut unmatched = diagnostics.into().iter().collect::<Vec<_>>();
+        for assertion in assertions.into() {
+            if !self.matches(assertion, &mut unmatched) {
+                failures.push(assertion.unmatched());
+            }
+        }
+        for diagnostic in unmatched {
+            failures.push(diagnostic.unmatched());
+        }
+        if failures.is_empty() {
+            Ok(())
+        } else {
+            Err(failures)
+        }
+    }
+
+    fn column<T: Diagnostic>(&self, diagnostic: &T) -> OneIndexed {
+        self.line_index
+            .source_location(diagnostic.range().start(), &self.source)
+            .column
+    }
+
+    /// Check if `assertion` matches any [`Diagnostic`]s in `unmatched`.
+    ///
+    /// If so, return `true` and remove the matched diagnostics from `unmatched`. Otherwise, return
+    /// `false`.
+    ///
+    /// An `Error` assertion can only match one diagnostic; even if it could match more than one,
+    /// we short-circuit after the first match.
+    ///
+    /// A `Type` assertion must match a revealed-type diagnostic, and may also match an
+    /// undefined-reveal diagnostic, if present.
+    fn matches<T: Diagnostic>(&self, assertion: &Assertion, unmatched: &mut Vec<&T>) -> bool {
+        let mut matched: FxHashSet<usize> = FxHashSet::default();
+        match assertion {
+            Assertion::Error(error) => {
+                for (index, diagnostic) in unmatched.iter().enumerate() {
+                    if !error.rule.is_some_and(|rule| rule != diagnostic.rule())
+                        && !error
+                            .column
+                            .is_some_and(|col| col != self.column(*diagnostic))
+                        && !error
+                            .message_contains
+                            .is_some_and(|needle| !diagnostic.message().contains(needle))
+                    {
+                        matched.insert(index);
+                        break;
+                    }
+                }
+            }
+            Assertion::Type(ta) => {
```

Comment:
Similar to @MichaReiser's nit elsewhere, I'd prefer `type_assertion`, `t_assertion`, or `ty` here rather than `ta` :-)